File: tests/suites/knowledge.js
// Phase 5 — Knowledge Workspace, notifications and administration.
//
// Curation being analyst-only is asserted deliberately: App Spec §3.23 denies
// Curate to the Facilitator, which looks like a bug but is separation of
// duties (§3.9's dedicated Knowledge Analyst role).

import { runSuite, clients, client, token, cleanup, resetDemoState, ACCOUNTS } from "../lib/harness.js";

runSuite("knowledge", async ({ baseUrl, results: r, c, DEMO_WORKSHOP: WS }) => {
  const { facilitator: F, analyst: A, participant: P } = await clients(baseUrl, ["facilitator", "analyst", "participant"]);

  const seeded = [];
  let assetId = null;

  try {
    r.section('knowledge search is scoped and registry-driven');
    const all = await F('GET', '/knowledge/search');
    r.ok(all.success, 'search returns results', all.error && all.error.message);
    const kinds = [...new Set((all.data || []).map(h => h.object_kind))];
    r.ok(kinds.length > 0, 'covers multiple object kinds', kinds.join(','));

    const q = await F('GET', '/knowledge/search?q=cloud');
    r.ok(q.success, 'full-text query works', q.success ? (q.data.length + ' hits') : q.error.message);

    const filtered = await F('GET', '/knowledge/search?object_type=factor');
    r.ok(filtered.success && filtered.data.every(h => h.object_kind === 'factor'),
      'object_type filter applies', filtered.data && filtered.data.length + ' factors');

    const byWorkshop = await F('GET', `/knowledge/search?workshop_id=${WS}`);
    r.ok(byWorkshop.success && byWorkshop.data.every(h => h.workshop_id === WS),
      'workshop filter applies');

    // The security property that matters: a non-member sees nothing from a
    // workspace they do not belong to.
    const orphan = (await c.query(
      `select id from public.profiles where id not in (select user_id from public.workspace_members) limit 1`)).rows[0];
    if (orphan) {
      const email = (await c.query('select email from public.profiles where id=$1', [orphan.id])).rows[0].email;
      const O = mk(await tok(email));
      const leak = await O('GET', '/knowledge/search');
      r.ok(leak.success && leak.data.length === 0,
        'a non-member of any workspace sees nothing', leak.data && leak.data.length);
    } else {
    r.note('skipped non-member check: every profile is a workspace member');
    }
    r.section('build a chain, then trace it');
    const factors = (await F('GET', `/workshops/${WS}/factors`)).data;
    const syn = await F('POST', `/workshops/${WS}/syntheses`,
      { title: 'P5 theme', description: 'seeded', evidence: { factor: [factors[0].id, factors[1].id] } });
    seeded.push(['syntheses', syn.data.id]);
    const ins = await F('POST', `/workshops/${WS}/insights`,
      { title: 'P5 insight', description: 'seeded', evidence: { synthesis: [syn.data.id] } });
    seeded.push(['insights', ins.data.id]);
    const rec = await F('POST', `/workshops/${WS}/recommendations`,
      { title: 'P5 recommendation', description: 'do it', fields: { priority: 'high' },
        evidence: { insight: [ins.data.id] } });
    seeded.push(['recommendations', rec.data.id]);

    const trace = await F('GET', `/knowledge/trace?object_type=recommendation&object_id=${rec.data.id}`);
    r.ok(trace.success, 'trace returns', trace.error && trace.error.message);
    r.ok(!!(trace.data.levels.insight && trace.data.levels.synthesis && trace.data.levels.factor),
      'walks recommendation -> insight -> theme -> factor (spec 11.26)',
      Object.keys(trace.data.levels).join(','));

    const ins2 = await F('POST', `/workshops/${WS}/insights`,
      { title: 'P5 sibling insight', description: 'shares a theme', evidence: { synthesis: [syn.data.id] } });
    seeded.push(['insights', ins2.data.id]);
    const related = await F('GET', `/knowledge/related?object_type=insight&object_id=${ins.data.id}`);
    r.ok(related.success && related.data.some(n => n.object_id === ins2.data.id),
      'related finds an insight sharing the same theme', related.data && related.data.length + ' related');
    r.section('promotion is deliberate (spec 8.31)');
    const tooEarly = await A('POST', '/knowledge/assets',
      { object_kind: 'insight', object_id: ins.data.id });
    r.ok(!tooEarly.success, 'unapproved output cannot be promoted', tooEarly.error && tooEarly.error.message);

    await F('POST', `/workshops/${WS}/insights/${ins.data.id}/review`, { action: 'approve' });

    const byFacilitator = await F('POST', '/knowledge/assets',
      { object_kind: 'insight', object_id: ins.data.id });
    r.ok(!byFacilitator.success && byFacilitator.error.code === 'FORBIDDEN',
      'a facilitator may NOT curate (spec 3.23)', byFacilitator.error && byFacilitator.error.message);

    const promoted = await A('POST', '/knowledge/assets',
      { object_kind: 'insight', object_id: ins.data.id, summary: 'Worth remembering', tags: ['digital'] });
    r.ok(promoted.success && promoted.data.state === 'candidate',
      'an analyst promotes it to candidate', promoted.error && promoted.error.message);
    assetId = promoted.data.id;

    const dupe = await A('POST', '/knowledge/assets',
      { object_kind: 'insight', object_id: ins.data.id });
    r.ok(!dupe.success, 'cannot promote the same thing twice', dupe.error && dupe.error.message);

    const published = await A('POST', `/knowledge/assets/${assetId}/publish`);
    r.ok(published.success && published.data.state === 'published', 'candidate becomes published asset');

    const assets = await F('GET', '/knowledge/assets?state=published');
    r.ok(assets.success && assets.data.some(a => a.id === assetId),
      'published assets are listed', assets.data && assets.data.length);

    const promotedSearch = await F('GET', '/knowledge/search?promoted=true');
    r.ok(promotedSearch.success && promotedSearch.data.every(h => h.promoted),
      'search can filter to promoted knowledge only', promotedSearch.data && promotedSearch.data.length);
    r.section('notifications come from real state transitions (spec 8.32)');
    // A participant creates a factor; the facilitator rejects it.
    const pf = await P('POST', `/workshops/${WS}/factors`,
      { category_key: 'strength', title: 'P5 participant factor' });
    const rej = await F('POST', `/workshops/${WS}/factors/${pf.data.id}/review`,
      { action: 'reject', note: 'Not specific enough' });
    r.ok(rej.success, 'facilitator rejects the participant factor');

    const pNotes = await P('GET', '/notifications');
    const mine = (pNotes.data || []).find(n => n.object_id === pf.data.id);
    r.ok(!!mine, 'the author was notified', mine && mine.title);
    r.ok(mine && mine.body === 'Not specific enough', 'the reviewer note carried through');

    const fNotes = await F('GET', '/notifications');
    r.ok(!(fNotes.data || []).some(n => n.object_id === pf.data.id),
      'the reviewer was NOT notified of their own action');

    const readIt = await P('POST', `/notifications/${mine.id}/read`);
    r.ok(readIt.success, 'recipient can mark it read');
    const stealRead = await F('POST', `/notifications/${mine.id}/read`);
    r.ok(!stealRead.success, 'someone else cannot mark it read', stealRead.error && stealRead.error.message);

    await c.query('delete from public.factors where id=$1', [pf.data.id]);
    r.section('administration is gated on global_role (spec 11.29)');
    const denied = await F('GET', '/admin/users');
    r.ok(!denied.success && denied.error.code === 'FORBIDDEN',
      'an ordinary user cannot list users', denied.error && denied.error.message);

    const jane = (await c.query(`select id from public.profiles where email='jane.smith@example.com'`)).rows[0].id;
    await c.query(`update public.profiles set global_role='platform_admin' where id=$1`, [jane]);
    const F2 = client(baseUrl, await token(ACCOUNTS.facilitator));

    const users = await F2('GET', '/admin/users');
    r.ok(users.success && users.data.length === 12, 'a platform admin lists users', users.data && users.data.length);

    const dana = (await c.query(`select id from public.profiles where email='dana.whitfield@example.com'`)).rows[0].id;
    const promote = await F2('PUT', `/admin/users/${dana}/role`, { global_role: 'admin' });
    r.ok(promote.success, 'admin can change a global role', promote.error && promote.error.message);
    await c.query(`update public.profiles set global_role='user' where id=$1`, [dana]);

    const selfDisable = await F2('POST', `/admin/users/${jane}/status`, { status: 'disabled' });
    r.ok(!selfDisable.success, 'an admin cannot disable themselves', selfDisable.error && selfDisable.error.message);

    const audit = await F2('GET', '/audit-events?object_type=knowledge_asset');
    r.ok(audit.success && audit.data.length > 0,
      'audit events are queryable and filtered', audit.data && audit.data.length);

    await c.query(`update public.profiles set global_role='user' where id=$1`, [jane]);

  } finally {
    r.section('cleanup');
    if (assetId) await c.query('delete from public.knowledge_assets where id=$1', [assetId]);
    await c.query('delete from public.knowledge_assets');
    for (const [t, id] of seeded.reverse()) await c.query(`delete from public.${t} where id=$1`, [id]);
    await c.query("delete from public.factors where title like 'P5 %'");
    await c.query("delete from public.notifications");
    await c.query("update public.profiles set global_role='user'");
    r.note('seeded rows, assets and notifications removed; roles reset');
    await c.end();
  }
});

File: pkg/handlers/knowledge.go
// The Knowledge Workspace (App Spec §2.12, §6.15, §8.31, §11.26).
//
// Organizational memory: search across every workshop the caller can see,
// follow an object's evidence chain, and deliberately promote the small
// number of outputs worth keeping.
//
// Everything is addressed through the object registry, so a sixth object kind
// becomes searchable, traceable and promotable without touching this file.
//
// The security rule that matters here (§12.24): a caller may only ever see
// objects in workspaces they belong to. Every query below is scoped by
// workspace membership, not filtered afterwards.
package handlers

import (
	"context"
	"encoding/json"
	"fmt"
	"net/http"
	"strings"
	"time"

	"github.com/go-chi/chi/v5"
	"github.com/jackc/pgx/v5"
	"github.com/jackc/pgx/v5/pgxpool"

	"swot-tows/pkg/audit"
	"swot-tows/pkg/authz"
	"swot-tows/pkg/httpctx"
	"swot-tows/pkg/objects"
	"swot-tows/pkg/response"
)

// Curators promote workshop output into organizational knowledge.
//
// App Spec §3.23's permission table gives Curate to the Analyst and NOT to
// the Facilitator. That looks surprising, but it is deliberate and matches
// §3.9's dedicated "Knowledge Analyst" role: curation is a separate
// discipline from running a workshop, and the person who ran it is not
// automatically the person who decides it is worth remembering. Followed as
// written rather than quietly widened.
var curatorRoles = []string{"analyst"}

// searchable is the set of kinds knowledge search covers, with the columns
// that carry their text. Reports are included per §6.15's search categories
// even though they are not an object-registry kind.
type searchableKind struct {
	Kind       string
	Table      string
	TitleCol   string
	BodyCol    string
	HasState   bool
}

func searchableKinds() []searchableKind {
	out := []searchableKind{
		{"factor", "factors", "title", "description", true},
	}
	for _, k := range objects.All() {
		body := "description"
		if !k.HasDescription {
			body = "narrative"
		}
		out = append(out, searchableKind{k.Key, k.Table, "title", body, true})
	}
	out = append(out, searchableKind{"report", "reports", "title", "", true})
	return out
}

type KnowledgeHit struct {
	ObjectKind   string  `json:"object_kind"`
	ObjectID     string  `json:"object_id"`
	Title        string  `json:"title"`
	Snippet      string  `json:"snippet"`
	State        string  `json:"state"`
	WorkshopID   string  `json:"workshop_id"`
	WorkshopName string  `json:"workshop_name"`
	CreatedAt    string  `json:"created_at"`
	Promoted     bool    `json:"promoted"`
	KnowledgeID  *string `json:"knowledge_id"`
}

// SearchKnowledge — GET /knowledge/search?q=&object_type=&workshop_id=&date_from=&date_to=
func SearchKnowledge(w http.ResponseWriter, r *http.Request) {
	r2, ok := httpctx.RequireAuth(w, r)
	if !ok {
		return
	}
	user := httpctx.UserFromContext(r2.Context())
	pool := mustPool(r2, w)
	if pool == nil {
		return
	}

	q := strings.TrimSpace(r2.URL.Query().Get("q"))
	kindFilter := r2.URL.Query().Get("object_type")
	workshopFilter := r2.URL.Query().Get("workshop_id")
	dateFrom := r2.URL.Query().Get("date_from")
	dateTo := r2.URL.Query().Get("date_to")
	promotedOnly := r2.URL.Query().Get("promoted") == "true"

	hits := []KnowledgeHit{}
	for _, k := range searchableKinds() {
		if kindFilter != "" && kindFilter != k.Kind {
			continue
		}

		body := "''"
		if k.BodyCol != "" {
			body = fmt.Sprintf("coalesce(o.%s, '')", k.BodyCol)
		}

		// Scoped by workspace membership in the JOIN, so an object the caller
		// cannot see never enters the result set to begin with.
		query := fmt.Sprintf(`
			select o.id::text, coalesce(o.%s, ''), %s, o.state, o.workshop_id::text, w.name, o.created_at,
			       ka.id::text
			from public.%s o
			join public.workshops w on w.id = o.workshop_id
			join public.workspace_members wsm
			     on wsm.workspace_id = w.workspace_id and wsm.user_id = $1
			left join public.knowledge_assets ka
			     on ka.object_kind = $2 and ka.object_id = o.id
			where true`, k.TitleCol, body, k.Table)

		args := []interface{}{user.ID, k.Kind}

		if q != "" {
			args = append(args, q)
			query += fmt.Sprintf(`
				and to_tsvector('english', coalesce(o.%s,'') || ' ' || %s)
				    @@ plainto_tsquery('english', $%d)`, k.TitleCol, body, len(args))
		}
		if workshopFilter != "" {
			args = append(args, workshopFilter)
			query += fmt.Sprintf(` and o.workshop_id = $%d`, len(args))
		}
		if dateFrom != "" {
			args = append(args, dateFrom)
			query += fmt.Sprintf(` and o.created_at >= $%d::timestamptz`, len(args))
		}
		if dateTo != "" {
			args = append(args, dateTo)
			query += fmt.Sprintf(` and o.created_at <= $%d::timestamptz`, len(args))
		}
		if promotedOnly {
			query += ` and ka.id is not null`
		}
		query += ` order by o.created_at desc limit 25`

		rows, err := pool.Query(r2.Context(), query, args...)
		if err != nil {
			response.Fail(w, response.CodeServerError, err.Error())
			return
		}
		for rows.Next() {
			var h KnowledgeHit
			var createdAt time.Time
			if err := rows.Scan(&h.ObjectID, &h.Title, &h.Snippet, &h.State,
				&h.WorkshopID, &h.WorkshopName, &createdAt, &h.KnowledgeID); err != nil {
				rows.Close()
				response.Fail(w, response.CodeServerError, err.Error())
				return
			}
			h.ObjectKind = k.Kind
			h.CreatedAt = createdAt.Format(time.RFC3339)
			h.Promoted = h.KnowledgeID != nil
			if len(h.Snippet) > 220 {
				h.Snippet = h.Snippet[:220] + "…"
			}
			hits = append(hits, h)
		}
		rows.Close()
	}

	response.OK(w, hits)
}

type TraceNode struct {
	ObjectKind string `json:"object_kind"`
	ObjectID   string `json:"object_id"`
	Title      string `json:"title"`
	State      string `json:"state"`
}

type TraceResult struct {
	Root TraceNode              `json:"root"`
	// Levels maps a cited kind to the objects at that level, walked through
	// the registry's own evidence links.
	Levels map[string][]TraceNode `json:"levels"`
}

// TraceKnowledge — GET /knowledge/trace?object_type=&object_id=
//
// App Spec §11.26 draws this as recommendation → insight → theme → artifact,
// but the shape comes from the registry's Evidence links, not from that list,
// so a methodology whose insights cite something else traces correctly too.
func TraceKnowledge(w http.ResponseWriter, r *http.Request) {
	r2, ok := httpctx.RequireAuth(w, r)
	if !ok {
		return
	}
	user := httpctx.UserFromContext(r2.Context())
	pool := mustPool(r2, w)
	if pool == nil {
		return
	}

	kindKey := r2.URL.Query().Get("object_type")
	objectID := r2.URL.Query().Get("object_id")
	if kindKey == "" || objectID == "" {
		response.Fail(w, response.CodeValidationError, "object_type and object_id are required")
		return
	}

	workshopID, title, state, errMsg := loadTraceRoot(r2.Context(), pool, kindKey, objectID)
	if errMsg != "" {
		response.Fail(w, response.CodeNotFound, errMsg)
		return
	}
	// Reuses the ordinary workshop membership check, so knowledge tracing
	// cannot be used to read across a boundary the rest of the app enforces.
	if _, err := authz.RequireWorkshopRole(r2.Context(), pool, user.ID, workshopID); err != nil {
		writeAuthzErr(w, err)
		return
	}

	out := TraceResult{
		Root:   TraceNode{ObjectKind: kindKey, ObjectID: objectID, Title: title, State: state},
		Levels: map[string][]TraceNode{},
	}

	kind, err := objects.ByKey(kindKey)
	if err == nil {
		supports := map[string][]SectionItem{}
		if err := expandEvidence(r2.Context(), pool, workshopID, kind, objectID, supports, 0); err != nil {
			response.Fail(w, response.CodeServerError, err.Error())
			return
		}
		for citeKind, items := range supports {
			nodes := []TraceNode{}
			for _, it := range items {
				nodes = append(nodes, TraceNode{
					ObjectKind: citeKind, ObjectID: it.ID, Title: it.Title, State: it.State,
				})
			}
			out.Levels[citeKind] = nodes
		}
	}

	response.OK(w, out)
}

func loadTraceRoot(ctx context.Context, pool *pgxpool.Pool, kindKey, objectID string) (
	workshopID, title, state, errMsg string,
) {
	table, _ := tableForKind(kindKey)
	if table == "" {
		if kindKey == "report" {
			table = "reports"
		} else {
			return "", "", "", "unknown object type"
		}
	}
	err := pool.QueryRow(ctx, fmt.Sprintf(
		`select workshop_id::text, coalesce(title, ''), state from public.%s where id = $1`, table),
		objectID).Scan(&workshopID, &title, &state)
	if err != nil {
		return "", "", "", "object not found"
	}
	return workshopID, title, state, ""
}

// RelatedKnowledge — GET /knowledge/related?object_type=&object_id=
//
// "Related" means sharing evidence: two insights that cite the same theme are
// related, which is a structural claim rather than a guess at similarity.
func RelatedKnowledge(w http.ResponseWriter, r *http.Request) {
	r2, ok := httpctx.RequireAuth(w, r)
	if !ok {
		return
	}
	user := httpctx.UserFromContext(r2.Context())
	pool := mustPool(r2, w)
	if pool == nil {
		return
	}

	kindKey := r2.URL.Query().Get("object_type")
	objectID := r2.URL.Query().Get("object_id")
	if kindKey == "" || objectID == "" {
		response.Fail(w, response.CodeValidationError, "object_type and object_id are required")
		return
	}

	kind, err := objects.ByKey(kindKey)
	if err != nil {
		response.OK(w, []TraceNode{})
		return
	}
	workshopID, _, _, errMsg := loadTraceRoot(r2.Context(), pool, kindKey, objectID)
	if errMsg != "" {
		response.Fail(w, response.CodeNotFound, errMsg)
		return
	}
	if _, err := authz.RequireWorkshopRole(r2.Context(), pool, user.ID, workshopID); err != nil {
		writeAuthzErr(w, err)
		return
	}

	seen := map[string]bool{objectID: true}
	out := []TraceNode{}

	// Siblings: other objects of the same kind citing any of the same things.
	for _, link := range kind.Evidence {
		rows, err := pool.Query(r2.Context(), fmt.Sprintf(`
			select distinct other.%s::text, coalesce(o.title,''), o.state
			from public.%s mine
			join public.%s other on other.%s = mine.%s and other.%s <> mine.%s
			join public.%s o on o.id = other.%s
			where mine.%s = $1
			limit 20`,
			link.SelfCol, link.Table, link.Table, link.OtherCol, link.OtherCol,
			link.SelfCol, link.SelfCol, kind.Table, link.SelfCol, link.SelfCol), objectID)
		if err != nil {
			continue
		}
		for rows.Next() {
			var n TraceNode
			if err := rows.Scan(&n.ObjectID, &n.Title, &n.State); err != nil {
				break
			}
			if seen[n.ObjectID] {
				continue
			}
			seen[n.ObjectID] = true
			n.ObjectKind = kindKey
			out = append(out, n)
		}
		rows.Close()
	}

	response.OK(w, out)
}

// ---- promotion ----

type KnowledgeAsset struct {
	ID           string   `json:"id"`
	WorkspaceID  string   `json:"workspace_id"`
	ObjectKind   string   `json:"object_kind"`
	ObjectID     string   `json:"object_id"`
	Title        string   `json:"title"`
	Summary      *string  `json:"summary"`
	State        string   `json:"state"`
	Tags         []string `json:"tags"`
	WorkshopName *string  `json:"workshop_name"`
	PromotedAt   string   `json:"promoted_at"`
	PublishedAt  *string  `json:"published_at"`
}

// ListKnowledgeAssets — GET /knowledge/assets?state=
func ListKnowledgeAssets(w http.ResponseWriter, r *http.Request) {
	r2, ok := httpctx.RequireAuth(w, r)
	if !ok {
		return
	}
	user := httpctx.UserFromContext(r2.Context())
	pool := mustPool(r2, w)
	if pool == nil {
		return
	}

	query := `
		select ka.id, ka.workspace_id, ka.object_kind, ka.object_id, ka.title, ka.summary,
		       ka.state, ka.tags, w.name, ka.promoted_at, ka.published_at
		from public.knowledge_assets ka
		join public.workspace_members wsm
		     on wsm.workspace_id = ka.workspace_id and wsm.user_id = $1
		left join public.workshops w on w.id = ka.source_workshop_id
		where true`
	args := []interface{}{user.ID}
	if st := r2.URL.Query().Get("state"); st != "" {
		args = append(args, st)
		query += fmt.Sprintf(` and ka.state = $%d`, len(args))
	}
	query += ` order by ka.promoted_at desc limit 100`

	rows, err := pool.Query(r2.Context(), query, args...)
	if err != nil {
		response.Fail(w, response.CodeServerError, err.Error())
		return
	}
	defer rows.Close()

	out := []KnowledgeAsset{}
	for rows.Next() {
		var a KnowledgeAsset
		var promotedAt time.Time
		var publishedAt *time.Time
		if err := rows.Scan(&a.ID, &a.WorkspaceID, &a.ObjectKind, &a.ObjectID, &a.Title,
			&a.Summary, &a.State, &a.Tags, &a.WorkshopName, &promotedAt, &publishedAt); err != nil {
			response.Fail(w, response.CodeServerError, err.Error())
			return
		}
		a.PromotedAt = promotedAt.Format(time.RFC3339)
		if publishedAt != nil {
			p := publishedAt.Format(time.RFC3339)
			a.PublishedAt = &p
		}
		out = append(out, a)
	}
	response.OK(w, out)
}

type promoteBody struct {
	ObjectKind string   `json:"object_kind"`
	ObjectID   string   `json:"object_id"`
	Summary    *string  `json:"summary"`
	Tags       []string `json:"tags"`
}

// PromoteKnowledge — POST /knowledge/assets
//
// §8.31: only reviewed output is worth remembering, so an object must be
// approved (or a report published) before it can become a candidate.
func PromoteKnowledge(w http.ResponseWriter, r *http.Request) {
	r2, ok := httpctx.RequireAuth(w, r)
	if !ok {
		return
	}
	user := httpctx.UserFromContext(r2.Context())
	pool := mustPool(r2, w)
	if pool == nil {
		return
	}

	var body promoteBody
	if err := json.NewDecoder(r2.Body).Decode(&body); err != nil {
		response.Fail(w, response.CodeValidationError, "invalid request body")
		return
	}

	workshopID, title, state, errMsg := loadTraceRoot(r2.Context(), pool, body.ObjectKind, body.ObjectID)
	if errMsg != "" {
		response.Fail(w, response.CodeNotFound, errMsg)
		return
	}

	// Curation is its own right (§3.23), checked against the source workshop.
	if _, err := authz.RequireWorkshopRole(r2.Context(), pool, user.ID, workshopID, curatorRoles...); err != nil {
		writeAuthzErr(w, err)
		return
	}

	if state != "approved" && state != "published" {
		response.Fail(w, response.CodeInvalidStateTransition,
			"Only approved output can become organizational knowledge.")
		return
	}

	var workspaceID string
	if err := pool.QueryRow(r2.Context(),
		`select workspace_id from public.workshops where id = $1`, workshopID).Scan(&workspaceID); err != nil {
		response.Fail(w, response.CodeServerError, err.Error())
		return
	}

	tags := body.Tags
	if tags == nil {
		tags = []string{}
	}

	var id string
	err := pool.QueryRow(r2.Context(), `
		insert into public.knowledge_assets
		  (workspace_id, source_workshop_id, object_kind, object_id, title, summary, state, promoted_by, tags)
		values ($1, $2, $3, $4, $5, $6, 'candidate', $7, $8)
		on conflict (object_kind, object_id) do nothing
		returning id`,
		workspaceID, workshopID, body.ObjectKind, body.ObjectID, title, trimPtr(body.Summary), user.ID, tags,
	).Scan(&id)
	if err == pgx.ErrNoRows {
		response.Fail(w, response.CodeValidationError, "This is already in the knowledge library.")
		return
	}
	if err != nil {
		response.Fail(w, response.CodeServerError, err.Error())
		return
	}

	audit.Record(r2.Context(), pool, user.ID, "knowledge.promoted", "knowledge_asset", id, "", "candidate",
		map[string]interface{}{"workshop_id": workshopID, "object_kind": body.ObjectKind})
	response.Created(w, map[string]string{"id": id, "state": "candidate"})
}

// PublishKnowledge — POST /knowledge/assets/{assetId}/publish
func PublishKnowledge(w http.ResponseWriter, r *http.Request) {
	r2, ok := httpctx.RequireAuth(w, r)
	if !ok {
		return
	}
	user := httpctx.UserFromContext(r2.Context())
	assetID := chi.URLParam(r2, "assetId")
	pool := mustPool(r2, w)
	if pool == nil {
		return
	}

	var workspaceID, state string
	var sourceWorkshop *string
	if err := pool.QueryRow(r2.Context(),
		`select workspace_id::text, state, source_workshop_id::text
		 from public.knowledge_assets where id = $1`, assetID,
	).Scan(&workspaceID, &state, &sourceWorkshop); err != nil {
		response.Fail(w, response.CodeNotFound, "knowledge asset not found")
		return
	}
	if sourceWorkshop != nil {
		if _, err := authz.RequireWorkshopRole(r2.Context(), pool, user.ID, *sourceWorkshop, curatorRoles...); err != nil {
			writeAuthzErr(w, err)
			return
		}
	}
	if state == "published" {
		response.Fail(w, response.CodeInvalidStateTransition, "This is already published.")
		return
	}

	if _, err := pool.Exec(r2.Context(), `
		update public.knowledge_assets
		set state = 'published', published_by = $2, published_at = now()
		where id = $1`, assetID, user.ID); err != nil {
		response.Fail(w, response.CodeServerError, err.Error())
		return
	}

	audit.Record(r2.Context(), pool, user.ID, "knowledge.published", "knowledge_asset", assetID,
		state, "published", map[string]interface{}{"workspace_id": workspaceID})
	response.OK(w, map[string]string{"id": assetID, "state": "published"})
}

// RemoveKnowledge — DELETE /knowledge/assets/{assetId}
func RemoveKnowledge(w http.ResponseWriter, r *http.Request) {
	r2, ok := httpctx.RequireAuth(w, r)
	if !ok {
		return
	}
	user := httpctx.UserFromContext(r2.Context())
	assetID := chi.URLParam(r2, "assetId")
	pool := mustPool(r2, w)
	if pool == nil {
		return
	}

	var sourceWorkshop *string
	if err := pool.QueryRow(r2.Context(),
		`select source_workshop_id::text from public.knowledge_assets where id = $1`, assetID,
	).Scan(&sourceWorkshop); err != nil {
		response.Fail(w, response.CodeNotFound, "knowledge asset not found")
		return
	}
	if sourceWorkshop != nil {
		if _, err := authz.RequireWorkshopRole(r2.Context(), pool, user.ID, *sourceWorkshop, curatorRoles...); err != nil {
			writeAuthzErr(w, err)
			return
		}
	}

	if _, err := pool.Exec(r2.Context(), `delete from public.knowledge_assets where id = $1`, assetID); err != nil {
		response.Fail(w, response.CodeServerError, err.Error())
		return
	}
	audit.Record(r2.Context(), pool, user.ID, "knowledge.removed", "knowledge_asset", assetID, "", "deleted", nil)
	response.OK(w, map[string]string{"id": assetID})
}
